src/api/graphapi.h

```diff
@@ -0,0 +1,417 @@
+
+RG_Info RG_TraversalDesc_out    // Traverse from given node on outgoing edges.
+(
+    RG_Node n,                  // Node to traverse from
+    const char *type,           // Type of out edges to follow
+    RG_Node *o                  // reachable node(s)
+);
+
+RG_Info RG_TraversalDesc_outE   // Traverse from given node on outgoing edges.
+(
+    RG_Node n,                  // Node to traverse from
+    const char *type,           // Type of out edges to follow
+    RG_Edge *e                  // reachable edge(s)
+);
+
+RG_Info RG_TraversalDesc_outV   // Extract source node from edge
+(
+    RG_Edge e,                  // Edge to get src node from    
+    RG_Node *n                  // Source node
+);
+
+//------------------------------------------------------------------------------
+// RG_TraversalDesc_in
+//------------------------------------------------------------------------------
+
+RG_Info RG_TraversalDesc_in    // Traverse given node on incoming edges.
+(
+    RG_Node n,                  // Node to reach
+    const char *type,           // Type of incoming edges to traverse
+    RG_Node *i                  // reachable node(s)
+);
+
+RG_Info RG_TraversalDesc_inE   // Traverse given node on incoming edges.
+(
+    RG_Node n,                  // Node to reach
+    const char *type,           // Type of incoming edges to traverse
+    RG_Edge *e                  // reachable edge(s)
+);
+
+RG_Info RG_TraversalDesc_inV    // Extract destination node from edge.
+(
+    RG_Edge e,                  // Edge to get dest node from
+    RG_Node *n                  // Destination node
+);
+
+//------------------------------------------------------------------------------
+// RG_Has
+//------------------------------------------------------------------------------
+
+RG_Info RG_Has_BOOL             // Entity must have given attribute set to value
+(
+    RG_Entity e,                // Graph entity to add filter to
+    const char *attribute,      // Attribute name
+    bool v                      // Attribute value
+);
+
+RG_Info RG_Has_INT64
+(
+    RG_Entity e,                // Graph entity to add filter to
+    const char *attribute,      // Attribute name
+    int64_t v
+);
+
+RG_Info RG_Has_UINT64
+(
+    RG_Entity e,                // Graph entity to add filter to
+    const char *attribute,      // Attribute name
+    uint64_t v                  // Attribute value
+);
+
+RG_Info RG_Has_FP32
+(
+    RG_Entity e,                // Graph entity to add filter to
+    const char *attribute,      // Attribute name
+    float v                     // Attribute value
+);
+
+RG_Info RG_Has_FP64
+(
+    RG_Entity e,                // Graph entity to add filter to
+    const char *attribute,      // Attribute name
+    double v                    // Attribute value
+);
+
+#define RG_Has(e,a,v)                               \
+    _Generic                                        \
```

Comment:
`_Generic` is a C11 introduction that C++ compilers don't handle, so this won't be compatible with our test suite.